contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+# TypeScript Coding Standards
+
+For our Javascript code style rules, refer to the [JavaScript Style Guide](STYLE.md). 
+
+You can also refer to [the TypeScript Handbook](https://www.typescriptlang.org/docs/handbook/intro.html) for more information on how to use TypeScript.
+
+## Organizing Type Definitions
+
+### File structure
+
+- Reusable type definitions, such as models (e.g. Report), must have their own file and be placed in a shared directory, e.g. `src/types/`.
+- Types specific to a single file should remain inside that file to keep their scope limited and their usage clear.
+- Types specific to a component/file with platform-specific variants should have their own file and be placed in the same folder.
+
+### File Extensions
+
+For TypeScript files containing JSX code, use the `.tsx` file extension. For other TypeScript code use the `.ts` file extension.
+
+Files containing only type declarations should **not** use the `.d.ts` extension. The only exception is when defining a common interface for platform-specific implementations (see below).
+
+### Platform-Specific Variants
+
+In most cases, the code written for this repo should be platform-independent. In such cases, each module should have a single file, `index.ts`, which defines the module's exports. There are, however, some cases in which a feature is intrinsically tied to the underlying platform. In such cases, the following file extensions can be used to export platform-specific code from a module:
+
+- `index.native.ts` - Native (iOS and Android)
+- `index.android.ts` - Android
+- `index.ios.ts` - iOS
+- `index.website.ts` - Web
+- `index.desktop.ts` - Desktop
+
+Note that `index.ts` should contain the default implementation, and only platform-specific implementations should be done in their respective files. i.e: If you have mobile-specific implementation in `index.native.ts`, then the desktop/web implementation can be contained in a shared `index.ts`.
+
+For each platform-specific module, create shared type definitions in a separate `types.ts` file and place it in the same folder. `types.ts` has to export shared types which are **compatible with all platform-specific implementations**. Declare component props, return types, and other common types in this file.
```

Comment:
Will `tsc` validate that each platform-specific implementation is compatible with the shared types? i.e: it needs to check types not only for `index.ts`, but also `index.native.ts`. Will `tsc` handle that for us?

Reply:
> Will tsc validate that each platform-specific implementation is compatible with the shared types? 

Shared types (types defined in `types.ts`) will be used in each variant. Each TS file is checked individually, meaning that if a shared type isn't compatible with implementation we will get an error (error will be showed in platform specific implementations, not in `types.ts` file).

> Will tsc validate that each platform-specific implementation is compatible with other platform specific implementations?

No, from TS perspective all these files are separate modules that aren't connected, that means there might be a case where `index.ios.ts` exports function named 'x' and `index.android.ts` exports function named 'xx' and tsc won't complain.

Reply:
> No, from TS perspective all these files are separate modules that aren't connected, that means there might be a case where index.ios.ts exports function named 'x' and index.android.ts exports function named 'xx' and tsc won't complain.

@roryabraham This is the reason we have to define shared types in `types.ts` and make sure you export them in each platform-specific file.